03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
-						assert(found);
-						//
-						dsLayouts[i] = nbl::core::move_and_dynamic_cast<IGPUDescriptorSetLayout>(found);
-					}
-				}
-				// The signature is `T* const& smart_refctd_ptr<T>::get()` which makes it possible to neatly
-				// get a raw pointer to a constant array of raw pointers from an array of smart pointers
-				// by taking the address of the first element's get() method return value
-				auto pool = m_device->createDescriptorPoolForDSLayouts(IDescriptorPool::ECF_NONE,&dsLayouts->get(),&dsLayouts->get()+MaxDescriptorSets);
-				pool->createDescriptorSets(MaxDescriptorSets,&dsLayouts->get(),ds);
-			}
+					.binding = 2,
+					.type = nbl::asset::IDescriptor::E_TYPE::ET_STORAGE_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE, // not is not the time for descriptor indexing
+					.stageFlags = IGPUShader::ESS_COMPUTE,
+					.count = 4, // TODO: check what will happen with: .count = 5
+					.samplers = nullptr // irrelevant for a buffer
+				},
+		};
+
+		nbl::video::IGPUDescriptorSetLayout::SBinding bindingsDS3[1] = {
+				{
+					.binding = 6,
+					.type = nbl::asset::IDescriptor::E_TYPE::ET_STORAGE_BUFFER,
+					.createFlags = IGPUDescriptorSetLayout::SBinding::E_CREATE_FLAGS::ECF_NONE,
+					.stageFlags = IGPUShader::ESS_COMPUTE,
+					.count = 1,
+					.samplers = nullptr // irrelevant for a buffer
+				},
+		};
+		
+		smart_refctd_ptr<IGPUDescriptorSetLayout> dsLayout1 = m_device->createDescriptorSetLayout(bindingsDS1);
+		smart_refctd_ptr<IGPUDescriptorSetLayout> dsLayout3 = m_device->createDescriptorSetLayout(bindingsDS3);
+		if (!dsLayout1 || !dsLayout3)
+			return logFail("Failed to create a Descriptor Layout!\n");
+
+		// Nabla actually has facilities for SPIR-V Reflection and "guessing" pipeline layouts for a given SPIR-V which we'll cover in a different example
+		smart_refctd_ptr<nbl::video::IGPUPipelineLayout> pplnLayout = m_device->createPipelineLayout({}, nullptr, smart_refctd_ptr(dsLayout1), nullptr,  smart_refctd_ptr(dsLayout3));
+		if (!pplnLayout)
+			return logFail("Failed to create a Pipeline Layout!\n");
+
+		// We use strong typing on the pipelines (Compute, Graphics, Mesh, RT), since there's no reason to polymorphically switch between different pipelines
+		smart_refctd_ptr<nbl::video::IGPUComputePipeline> pipeline;
+		{
+			IGPUComputePipeline::SCreationParams params = {};
+			params.layout = pplnLayout.get();
+			// Theoretically a blob of SPIR-V can contain multiple named entry points and one has to be chosen, in practice most compilers only support outputting one (and glslang used to require it be called "main")
+			params.shader.entryPoint = "main";
+			params.shader.shader = shader.get();
+			// we'll cover the specialization constant API in another example
+			if (!m_device->createComputePipelines(nullptr, { &params,1 }, &pipeline))
+				return logFail("Failed to create pipelines (compile & link shaders)!\n");
+		}
```

Comment:
we should probably test-drive the instrospector here, and ignore
```
// Nabla actually has facilities for SPIR-V Reflection and "guessing" pipeline layouts for a given SPIR-V which we'll cover in a different example
```

the one thing you'll need to hand-roll here is creation of IGPUComputePipeline and IGPUDescriptorSet from ICPUComputePipeline and ICPUDescriptorSet along with all dependants (layouts and shader), without resorting to the asset converter (I wanna do that in example 09 or 24/25).